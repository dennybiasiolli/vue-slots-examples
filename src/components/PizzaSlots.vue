<script setup lang="ts">
import { CrustComposition, CrustStyle, Sauce, Topping } from '@/types'

export interface Props {
  name?: string
  crustStyle: {
    key: CrustStyle
    name: string
    size: string
  }
  crustComposition: {
    key: CrustComposition
    name: string
    color: string
  }
  sauce: {
    key: Sauce
    name: string
    color: string
  }
  toppings: {
    key: Topping
    name: string
  }[]
}

defineProps<Props>()
</script>

<template>
  <div class="pizza-box">
    <h1 class="pizza-title">{{ name }}</h1>
    <slot name="pizza">
      <slot name="crust">
        <div>{{ crustStyle.name }}, made with {{ crustComposition.name.toLowerCase() }}</div>
      </slot>

      <slot name="sauce" :sauce="sauce">
        <div>{{ sauce.name }}</div>
      </slot>

      <slot name="toppings">
        Toppings:
        <ul>
          <li v-for="topping in toppings" :key="topping.key">
            <slot name="topping" :topping="topping">
              {{ topping.name }}
            </slot>
          </li>
        </ul>
      </slot>
    </slot>
  </div>
</template>

<style scoped>
.pizza-box {
  border: 1px solid lightgrey;
  padding: 5px;
  width: 450px;
}
.pizza-title {
  font-weight: bold;
}
</style>
