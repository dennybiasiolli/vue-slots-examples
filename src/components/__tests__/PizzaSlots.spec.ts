import { describe, it, expect } from 'vitest'

import { mount } from '@vue/test-utils'
import PizzaSlots from '../PizzaSlots.vue'
import {
  CrustComposition,
  CrustStyle,
  Sauce,
  Topping,
  crustCompositionMap,
  crustStyleMap,
  sauceMap,
  toppingMap
} from '@/types'

describe('PizzaSlots', () => {
  it('with default content', () => {
    const wrapper = mount(PizzaSlots, {
      props: {
        name: 'Custom pizza',
        crustStyle: crustStyleMap[CrustStyle.PAN],
        crustComposition: crustCompositionMap[CrustComposition.WHOLE],
        sauce: sauceMap[Sauce.TOMATO],
        toppings: [toppingMap[Topping.MOZZARELLA], toppingMap[Topping.BASIL]]
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('h1.pizza-title').text()).toBe('Custom pizza')
  })

  it('with custom pizza slot', () => {
    const wrapper = mount(PizzaSlots, {
      props: {
        name: 'Custom pizza',
        crustStyle: crustStyleMap[CrustStyle.PAN],
        crustComposition: crustCompositionMap[CrustComposition.WHOLE],
        sauce: sauceMap[Sauce.TOMATO],
        toppings: [toppingMap[Topping.MOZZARELLA], toppingMap[Topping.BASIL]]
      },
      slots: {
        pizza: "<h2>foobar</h2>",
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('h1.pizza-title').text()).toBe('Custom pizza')
    expect(wrapper.find('h2').text()).toBe('foobar')
  })

  it('with custom topping scoped slot', () => {
    const wrapper = mount(PizzaSlots, {
      props: {
        name: 'Custom pizza',
        crustStyle: crustStyleMap[CrustStyle.PAN],
        crustComposition: crustCompositionMap[CrustComposition.WHOLE],
        sauce: sauceMap[Sauce.TOMATO],
        toppings: [toppingMap[Topping.MOZZARELLA], toppingMap[Topping.BASIL]]
      },
      slots: {
        topping: `<template #topping="{ topping }">abc-{{ topping.key }}</template>`,
        // also usable as
        // topping: 'abc-{{ params.topping.key }}',
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('h1.pizza-title').text()).toBe('Custom pizza')
    expect(wrapper.findAll('li')[0].text()).toBe('abc-mozzarella')
    expect(wrapper.findAll('li')[1].text()).toBe('abc-basil')
  })
})
