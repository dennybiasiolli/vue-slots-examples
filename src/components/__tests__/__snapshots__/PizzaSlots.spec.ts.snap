// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PizzaSlots > with custom pizza slot 1`] = `
"<div data-v-81b24109=\\"\\" class=\\"pizza-box\\">
  <h1 data-v-81b24109=\\"\\" class=\\"pizza-title\\">Custom pizza</h1>
  <h2>foobar</h2>
</div>"
`;

exports[`PizzaSlots > with custom topping scoped slot 1`] = `
"<div data-v-81b24109=\\"\\" class=\\"pizza-box\\">
  <h1 data-v-81b24109=\\"\\" class=\\"pizza-title\\">Custom pizza</h1>
  <div data-v-81b24109=\\"\\">Pan crust thicker and fluffer, made with whole wheat flour</div>
  <div data-v-81b24109=\\"\\">Tomato sauce</div> Toppings: <ul data-v-81b24109=\\"\\">
    <li data-v-81b24109=\\"\\">abc-mozzarella</li>
    <li data-v-81b24109=\\"\\">abc-basil</li>
  </ul>
</div>"
`;

exports[`PizzaSlots > with default content 1`] = `
"<div data-v-81b24109=\\"\\" class=\\"pizza-box\\">
  <h1 data-v-81b24109=\\"\\" class=\\"pizza-title\\">Custom pizza</h1>
  <div data-v-81b24109=\\"\\">Pan crust thicker and fluffer, made with whole wheat flour</div>
  <div data-v-81b24109=\\"\\">Tomato sauce</div> Toppings: <ul data-v-81b24109=\\"\\">
    <li data-v-81b24109=\\"\\">Mozzarella</li>
    <li data-v-81b24109=\\"\\">Basil</li>
  </ul>
</div>"
`;
